### PR TITLE
feat(api): capture mouseover event on a track; click event on a linear track

### PR DESCRIPTION
**Pull request by sehilyi**
```ts
goslingRef.current.api.subscribe('trackMouseOver', (type, e) ⇒ {
  console.log(
    type, // 'trackMouseOver'
    e.id, // identical to `id` (random ID when unspecified), e.g., "track-2"
    e.spec, // spec of a source track, e.g., { assembly: 'hg38', layout: 'linear', ...}
    e.shape, // shape of the track, e.g., {x: 0, y: 250, width: 600, height: 130}
  )
});
```

```ts
goslingRef.current.api.subscribe('trackClick', (type, e) ⇒ {
  console.log(
    type, // 'trackClick'
    e.id, // identical to `id` (random ID when unspecified), e.g., "track-2"
    e.spec, // spec of a source track, e.g., { assembly: 'hg38', layout: 'linear', ...}
    e.shape, // shape of the track, e.g., {x: 0, y: 250, width: 600, height: 130}
  )
});
```

(We only support Click event on linear tracks.)

### CODE REVIEW

File: schema/gosling.schema.json [hunk 1]
```diff
@@ -448,14 +448,9 @@
...
           "description": "Whether to use offset of the domain proportionally. This is bound to brushes on the color legend. __Default__: `[0, 1]`",
-          "items": [
-            {
-              "type": "number"
-            },
-            {
-              "type": "number"
-            }
-          ],
+          "items": {
```
Comment (sehilyi) [inline schema/gosling.schema.json]: Changes in `schema.json` files seem to be due to the upgrade of `ts-json-schema-generator`

**Comment by sehilyi** [comment]
Not sure if the name `trackMouseOver` is fine.

We can later probably unify the track-level events with mark-level events (`mouseOver`, `rangeSelect`, `click`).

File: package.json [hunk 2]
```diff
@@ -131,7 +131,7 @@
         "react-split-pane": "^0.1.91",
         "remark-gfm": "^1.0.0",
         "strip-json-comments": "^3.1.1",
-        "ts-json-schema-generator": "^0.75.0",
```
Comment (manzt) [inline package.json:134]: Was this just routine maintenance  or was there something in this PR that required updating the schema generator?

File: src/core/gosling.schema.ts [hunk 3]
```diff
@@ -254,16 +254,51 @@ interface RangeMouseEventData extends CommonEventData {
     genomicRange: [GenomicPosition, GenomicPosition] | null;
 }
 
+/** The visual parameters that determine the shape of a linear track */
+interface LinearTrackShape {
+    x: number; // The origin is the left top corner
+    y: number;
+    width: number;
+    height: number;
+}
```
Comment (manzt) [inline src/core/gosling.schema.ts]: Maybe move this comment so that it is picked up by JSDoc.

```suggestion
/** 
 * The visual parameters that determine the shape of a linear track. 
 * Origin is the left top corner.
 */
interface LinearTrackShape {
    x: number;
    y: number;
    width: number;
    height: number;
}
```

File: src/core/gosling.schema.ts [hunk 4]
```diff
@@ -254,16 +254,51 @@ interface RangeMouseEventData extends CommonEventData {
...
+
+/** The visual parameters that determine the shape of a circular track */
+interface CircularTrackShape {
+    cx: number;
+    cy: number;
+    innerRadius: number;
+    outerRadius: number;
+    // Anticlockwise, starts from 12 o'clock.
+    startAngle: number; // [0, 360]
+    endAngle: number; // [0, 360]
```
Comment (manzt) [inline src/core/gosling.schema.ts]: Each property can be annotated above with `/** ... */` and it will add to the documentation.

File: src/core/gosling.schema.ts [hunk 5]
```diff
@@ -254,16 +254,51 @@ interface RangeMouseEventData extends CommonEventData {
...
+}
+
+/** The informatino for a mouse event on a track */
```
Comment (manzt) [inline src/core/gosling.schema.ts]:
```suggestion
/** The information for a track mouse event */
```

File: src/core/gosling.schema.ts [hunk 6]
```diff
@@ -254,16 +254,51 @@ interface RangeMouseEventData extends CommonEventData {
...
+type TrackMouseEventData = {
+    /** ID of a source track, i.e., `track.id` */
+    id: string;
+
+    /** Source spec that is processed by the Gosling compiler, e.g., default properties filled in. */
```
Comment (manzt) [inline src/core/gosling.schema.ts]:
```suggestion
    /** Expanded track specification processed by the Gosling compiler, e.g., default properties filled in. */
```

File: src/core/gosling.schema.ts [hunk 7]
```diff
@@ -254,16 +254,51 @@ interface RangeMouseEventData extends CommonEventData {
...
 export type _EventMap = {
     mouseOver: PointMouseEventData;
     click: PointMouseEventData;
     rangeSelect: RangeMouseEventData;
     rawData: CommonEventData;
+    trackMouseOver: TrackMouseEventData;
 };
 
 export type MouseEventsDeep = {
+    // TODO(Jul-19-2022): Due to cyclic definition of `SingleTrack`, `ts-json-schema-generator` throws an error
```
Comment (manzt) [inline src/core/gosling.schema.ts]: ah I see now why the change was necessary

File: src/core/gosling.schema.ts [hunk 8]
```diff
@@ -254,16 +254,51 @@ interface RangeMouseEventData extends CommonEventData {
...
+    // see https://github.com/vega/ts-json-schema-generator/issues/192
+    // [Event in keyof Omit<_EventMap, 'rawData'>]?: boolean;
     /** Turn on and off individual mouse events. */
-    [Event in keyof Omit<_EventMap, 'rawData'>]?: boolean;
+    [Event in 'mouseOver' | 'click' | 'rangeSelect']?: boolean;
```
Comment (manzt) [inline src/core/gosling.schema.ts]: It might be easier if we just specified these normally.

```typescript
export interface MouseEventsDeep {
    /** Whether to enable mouseover events. */
    mouseOver: boolean;
    /** Whether to enable click events. */
    click: boolean;
    /** Whether to send range selection events. */
    rangeSelect: boolean;
    /** Group marks using keys in a data field. This affects how a set of marks are highlighted/selected by interaction. __Default__: `undefined` */
    groupMarksByField?: string;
    /** Determine whether all marks underneath the mouse point should be affected by mouse over. __Default__: `false` */
    enableMouseOverOnMultipleMarks?: boolean;
}
```
the original motivation was to reuse the `_EventMap` object, but if we can't then I think it's fine to just have a normal type definition.

File: src/core/utils/polar.ts [hunk 9]
```diff
@@ -29,3 +29,10 @@ export function positionToRadian(x: number, y: number, cx: number, cy: number) {
         return Math.atan((y - cy) / (x - cx)) - Math.PI;
     }
 }
+
+/**
+ * Calculate a degree in the range of [0, 360) based on two points. Anticlockwise, starts from 12 o'clock.
+ */
+export function pointsToDegree(x: number, y: number, cx: number, cy: number) {
+    return ((Math.atan2(-(y - cy), x - cx) / Math.PI) * 180 + 270) % 360;
+}
```
Comment (manzt) [inline src/core/utils/polar.ts:38]: Maybe add a couple unit tests?

File: src/gosling-mouse-event/polygon.ts [hunk 10]
```diff
@@ -1,3 +1,31 @@
+import { pointsToDegree } from '../core/utils/polar';
+
+/**
+ * @param {Array} point Tuple of the form `[x,y]` to be tested.
```
Comment (manzt) [inline src/gosling-mouse-event/polygon.ts]: It is usually preferred when using TS with JSDoc to omit the type definition (within `{ }`) in the JSDoc comment. e..,g

```
 * @param point Tuple of the form `[x,y]` to be tested.
 ```

This is because the TypeScript type is much more precise, and the JSDoc comment is ignored by the typechecker completely. We have some JSDoc comments like this in Viv, but that's because our documentation generator relies on the type in the JSDoc (and I'd prefer to get rid of it).

File: src/gosling-track/gosling-track.ts [hunk 11]
```diff
@@ -84,6 +85,8 @@ function GoslingTrack(HGC: import('@higlass/types').HGC, ...args: any[]): any {
         private mRangeBrush: LinearBrushModel;
         private _xScale!: ScaleLinear<number, number>;
         private _yScale!: ScaleLinear<number, number>;
+        private position!: [number, number]; // left and top offsets of a track
+        private dimensions!: [number, number]; // width and height of a track
```
Comment (manzt) [inline src/gosling-track/gosling-track.ts]: Changing these to JSDoc comments means the information is accessible when hovering the property anywhere in the file:

```suggestion
        /** left and top offsets of a track */
        private position!: [number, number];
        /** width and height of a track */
        private dimensions!: [number, number];
```

image

File: src/gosling-track/gosling-track.ts [hunk 12]
```diff
@@ -1233,6 +1236,39 @@ function GoslingTrack(HGC: import('@higlass/types').HGC, ...args: any[]): any {
...
+                    isPointInsideDonutSlice(
+                        [mouseX, mouseY],
+                        [width / 2.0, height / 2.0],
+                        [innerRadius, outerRadius],
+                        [startAngle, endAngle]
+                    )
+                ) {
+                    publish('trackMouseOver', {
+                        id: this.viewUid,
+                        spec: JSON.parse(JSON.stringify(this.options.spec)),
```
Comment (manzt) [inline src/gosling-track/gosling-track.ts]: I believe you could just use the newish [`structuredClone`](https://developer.mozilla.org/en-US/docs/Web/API/structuredClone) method here:

```suggestion
                        spec: structuredClone(this.options.spec),
```

File: src/gosling-track/gosling-track.ts [hunk 13]
```diff
@@ -1233,6 +1236,39 @@ function GoslingTrack(HGC: import('@higlass/types').HGC, ...args: any[]): any {
...
+                        shape: { cx, cy, innerRadius, outerRadius, startAngle, endAngle }
+                    });
+                }
+            } else {
+                publish('trackMouseOver', {
+                    id: this.viewUid,
+                    spec: JSON.parse(JSON.stringify(this.options.spec)),
```
Comment (manzt) [inline src/gosling-track/gosling-track.ts]: Likewise here re: structuredClone

**Review by manzt — APPROVED** [review]
Nice work! Just some minor comments. I don't feel too strongly about the naming, but I could imagine some type of grouping by event type that you can configure a listener for:

```typescript
export type MouseEventsDeep = {
   mouseOver?: "track" | "point" | "all"
```

and then the events could be prefixed: `track.mouseOver`

Not sure if I like that better than what we have, but just an idea.

File: package.json (hunk 2 above)
Comment (sehilyi) [reply package.json:134]: It is just routine maintenance and is not specific to new features introduced in this PR.

**Comment by sehilyi** [comment]
Thank you for all feedback! FYI, I added support for `click` on a linear track as well with small code modifications.

> I don't feel too strongly about the naming, but I could imagine some type of grouping by event type that you can configure a listener for:

I was also thinking the similar approach but not sure as well what is the best approach. I think we need to build a big picture regarding what mouse events we want to support.
